doc/developer/design/20230714_optimizer_interface.md

```diff
@@ -0,0 +1,336 @@
+# <Insert Name>
+
+- Associated: (Insert list of associated epics, issues, or PRs)
+
+<!--
+The goal of a design document is to thoroughly discover problems and
+examine potential solutions before moving into the delivery phase of
+a project. In order to be ready to share, a design document must address
+the questions in each of the following sections. Any additional content
+is at the discretion of the author.
+
+Note: Feel free to add or remove sections as needed. However, most design
+docs should at least keep the suggested sections.
+-->
+
+## The Problem
+
+<!--
+What is the user problem we want to solve?
+
+The answer to this question should link to at least one open GitHub
+issue describing the problem.
+-->
+
+Currently, Materialize executes optimization pipelines in the following situations:
+
+1. In `Coordinator::bootstrap`, when re-installing `MaterializedView` and`Index`
+   catalog items.
+1. In `Coordinator::sequence_plan`, when sequencing statements that result in an
+   executable dataflow or a `View` creation.
+1. In `Coordinator::sequence_plan`, when explaining statements that result in an
+   executable dataflow (`SELECT`, `CREATE MATERIALIZED VIEW`, or `CREATE
+   INDEX`).
+
+The optimization pipeline that we run nowadays consists of multiple IRs,
+lowerings (transformations between adjacent IRs), and optimizations (rewrites
+within the same IR). Because these steps have evolved slowly over time, they
+have never been unified under a single API.
+
+Consequently, all client code the `Coordinator` that runs an optimizer pipeline
+currently constructs this pipeline implicitly number of at-hoc API calls. This
+results in redundant or — worse still — _near_ redundant code.
+
+Keeping the optimization pipelines for each statement type `T` in sync across
+its client sites ("bootstrap `T`", "execute `T`", and "explain `T`") therefore
+is constantly at risk of getting out of sync. When this happens, it negatively
+impacts the overall product experience. For example, in the past we had
+situations with "`EXPLAIN` drift", where "explain `SELECT`" results wre not
```

Comment:
BTW this is the case even now because "execute `SELECT`" statements produce a dataflow with an exported `IndexDesc` that we don't create in the "explain `SELECT`". It's not something that currently affects the `EXPLAIN` output, but still...